download/download.go

```diff
@@ -159,44 +168,86 @@ func (d *Downloader) loop(ctx context.Context) {
 
 	var retry int
 
+	// Indicates if the loop should wait. When the triggering mode is manual,
+	// the downloader should not start automatically.
+	wait := *d.config.Trigger == TriggerManual
+	var done chan struct{}
+
 	for {
-		longPoll, err := d.oneShot(ctx)
+
+		var longPoll bool
+		var err error
+
+		if !wait {
+			longPoll, err = d.oneShot(ctx)
+
+			if *d.config.Trigger == TriggerManual {
+				close(done)
+				wait = true
+			}
+		}
 
 		if ctx.Err() != nil {
 			return
 		}
 
-		var delay time.Duration
+		var waitC chan chan struct{}
+
+		switch *d.config.Trigger {
+		case TriggerPolling:
+			var delay time.Duration
 
-		if err == nil {
-			if !longPoll {
-				if d.config.Polling.LongPollingTimeoutSeconds != nil {
-					d.config.Polling.LongPollingTimeoutSeconds = nil
+			if err == nil {
+				if !longPoll {
+					if d.config.Polling.LongPollingTimeoutSeconds != nil {
+						d.config.Polling.LongPollingTimeoutSeconds = nil
+					}
+
+					// revert the response header timeout value on the http client's transport
+					if *d.client.Config().ResponseHeaderTimeoutSeconds == 0 {
+						d.client = d.client.SetResponseHeaderTimeout(&d.respHdrTimeoutSec)
+					}
+
+					min := float64(*d.config.Polling.MinDelaySeconds)
+					max := float64(*d.config.Polling.MaxDelaySeconds)
+					delay = time.Duration(((max - min) * rand.Float64()) + min)
 				}
+			} else {
+				delay = util.DefaultBackoff(float64(minRetryDelay), float64(*d.config.Polling.MaxDelaySeconds), retry)
+			}
 
-				// revert the response header timeout value on the http client's transport
-				if *d.client.Config().ResponseHeaderTimeoutSeconds == 0 {
-					d.client = d.client.SetResponseHeaderTimeout(&d.respHdrTimeoutSec)
+			d.logger.Debug("Waiting %v before next download/retry.", delay)
+			timer := time.NewTimer(delay)
+
+			waitC = make(chan chan struct{}, 1)
+			go func() {
+				select {
+				case <-timer.C:
+					if err != nil {
+						retry++
+					} else {
+						retry = 0
+					}
+
+					if ctx.Err() != nil {
+						return
+					}
+
+					close(waitC)
+				case <-ctx.Done():
+					return
 				}
+			}()
 
-				min := float64(*d.config.Polling.MinDelaySeconds)
-				max := float64(*d.config.Polling.MaxDelaySeconds)
-				delay = time.Duration(((max - min) * rand.Float64()) + min)
-			}
-		} else {
-			delay = util.DefaultBackoff(float64(minRetryDelay), float64(*d.config.Polling.MaxDelaySeconds), retry)
+		case TriggerManual:
+			d.logger.Debug("Waiting for manual trigger before next download/retry.")
+			waitC = d.trigger
 		}
 
-		d.logger.Debug("Waiting %v before next download/retry.", delay)
-		timer := time.NewTimer(delay)
-
 		select {
-		case <-timer.C:
-			if err != nil {
-				retry++
-			} else {
-				retry = 0
-			}
+		case done = <-waitC:
```

Comment:
Is `done` used? 👀

Reply:
`done` is closed in the manual trigger case above to indicate to the caller that the plugin's work is finished.